js-packages/@prestojs/util/src/useAsync.ts

```diff
@@ -0,0 +1,439 @@
+
+    // =========================================================================
+    // This is the function that gets called to fire the async function. It's
+    // either called manually by the consumer calling the returned 'run' function
+    // or internally on mount and on any subsequent args changed when trigger
+    // is not MANUAL.
+    // Note that this function is never recreated - it's cached once with
+    // useCallback and from that point on should have consistent identity.
+    // It either gets passed the arguments or reads then from a ref at call time.
+    const execute = useCallback((
+        executeFn,
+        executeArgs,
+        // When execute is called read the current value for onSuccess and
+        // onError. This allows for nicer ergonomics by allowing inline
+        // arrow functions to be passed without triggering the function
+        // when trigger is not MANUAL. Without doing it this way it would
+        // need to be a dependency on the initial fetch useEffect below and
+        // as such would trigger a re-execute whenever it changed.
+        executeOnSuccess = lastValues.current.onSuccess,
+        executeOnError = lastValues.current.onError
+    ): [Promise<any>, () => void] => {
+        let isCurrent = true;
+        dispatch({ type: 'start' });
+        const promise = executeFn(...executeArgs);
+        if (!isPromise(promise)) {
+            const message = 'useAsync can only be used with functions that return a Promise. Got: ';
+            // eslint-disable-next-line no-console
+            console.error(message, promise);
+            throw new Error(message);
+        }
+        promise.then(
+            response => {
+                if (!isCurrent) {
+                    return;
+                }
+                dispatch({ type: 'success', payload: response });
+                if (executeOnSuccess) {
+                    executeOnSuccess(response);
+                }
+            },
+            e => {
+                if (!isCurrent) {
+                    return;
+                }
+                dispatch({ type: 'error', payload: e });
+                if (executeOnError) {
+                    executeOnError(e);
+                }
+            }
+        );
+        return [
+            promise,
+            (): void => {
+                isCurrent = false;
+            },
+        ];
+    }, []);
+    // =========================================================================
+
+    // abortRef is used by execute to store an abort function so that if the
+    // component unmounts before the async function finishes running then it
+    // can avoid calling success / error functions that would set state.
+    const abortRef = useRef<(() => void) | null>();
+
+    // =========================================================================
+    // Effect to trigger the execute method when:
+    // 1) Parameters change
+    // 2) The actual function changes (?)
```

Comment:
What's the `(?)` for?

Reply:
heh no idea too long ago... copied from djrad

Reply:
Is this because the function changing doesn't actually trigger a rerender?